Add ring support for 1D EDM Fabric

tt-aho opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/18798

### Problem description
We need to enable Ring support for 1D fabric to allow for better performance by reducing the number of hops needed. Ex on a T3K configured as an 8 device line, the max number of hops is 7. If it was an 8 device ring, this would be max 4 hops.

### What's changed
Add ring support for 1D EDM fabric.
This adds a new VC, which adds an additional sender receiver channel. Requires reducing the original number of sender/receiver buffer slots in half to support the new channels due to us currently using power of 2 buffer slots.
Large perf drop on Line was observed due to shrinking channel buffer slots, so add a flag to selectively enable ring support for testing. 
Still has an observed perf drop even though ring buffering code should be disabled, investigating.

Current perf numbers:
Line Mcast: ~6.7 B/c (unchanged)
Line Unicast: ~8.4 B/c (unchanged)
Ring Mcast: ~7.6 B/c

TODO:

- Try supporting non-power of 2 buffer slots to enable more buffering for ring buffer (not in this PR)
- Pipe the arg to control when to switch to high VC. We need to determine how we should figure out the optimal switchover point and pipe it into the user kernel to set up the packet header.
- Add more testing of ring support for functionality, and add a test that does deadlock without the second VC, but is resolved with using the second VC.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/13835457552
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

T3K Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/13835462915
T3K Fast: https://github.com/tenstorrent/tt-metal/actions/runs/13835572534
TG Unit: https://github.com/tenstorrent/tt-metal/actions/runs/13835587838
Microbench: https://github.com/tenstorrent/tt-metal/actions/runs/13835592831

tt-aho committed:
#0: Improve EDM packet header encoding logic

tt-aho committed:
#18798: Add support for encoding VC in EDM LowLatencyPacketHeader

tt-aho committed:
#18798: Add support for using 2 VCs in EDM host setup and kernel code

tt-aho committed:
#18798: Refactor EDM host code to use arrays for sender receiver channels
This makes it a lot easier to keep the channel logic consistent and reduce errors when duplicating channel logic

tt-aho committed:
#18798: Add support for disabling ring. Changes required for ring support was found to sustantially drop performance of line.
Update edm benchmark test to support ring

tt-aho committed:
#18798: Alternate which direction has longest hops in ring mcast perf test to better distribute VC traffic

tt-aho committed:
#18798: Change WriteTransactionIdTracker back to be more compile time based
This brings Line perf back up to close to what it was before Ring support was added

tt-aho committed:
#18798: Make VC switchover implicitly be 1/2 the send length for now
Future TODO is to set up/determine the switchover point from host

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -263,9 +260,9 @@ struct TransactionIdCounter {
 };
 
-template <size_t NUM_CHANNELS, size_t MAX_TRANSACTION_IDS>
+template <size_t NUM_CHANNELS, size_t MAX_TRANSACTION_IDS, size_t OFFSET>
```
Needed now because there are 2 receiver channels and we don't want them to have transaction IDs that collide

tt-aho replied:
Yes, that's correct. I had originally made the offset a runtime parameter to make this struct easy to put in an array for the two receivers, but that had a large negative impact on performance so reverted them to declare separate templated structs for now.

tt-aho commented on ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder.cpp:
```diff
@@ -41,97 +41,245 @@ namespace ttnn::ccl {
 //                         ------------------
 //
 
+FabricEriscDatamoverConfig::FabricEriscDatamoverConfig() {
```
Yeah. This was mostly a refactoring change to prevent easy to make errors from copy pasting when adding/modifying channel values (as I ended up making a number of mistakes 😞. Instead of defining each channel separately as a separate member, and then copy pasting all of the setup per channel, I've moved the sender and receiver channel members to all be arrays and initialized with for loops for consistent behaviour for a given member.
As a result, the current setup initialization code that was in the struct hpp wasn't viable because it initializes the members one by one, so I moved it to the cpp file and initialize the members through for loops. This makes the address/offset calculations easier to see and track for each member. The changes are tracked in separate commits. First 3 commits are the actual kernel/host code changes to add the VC, and the 4th commit is the refactoring. I can drop this is needed.

tt-aho commented on tests/tt_metal/microbenchmarks/ethernet/test_fabric_edm_bandwidth.py:
```diff
@@ -91,12 +101,17 @@ def run_fabric_edm(
-    "expected_bw",
-    [6.7],
-)
+@pytest.mark.parametrize("ring_topology, expected_bw", [(True, 7.3), (False, 6.4)])
```
I agree. I think I'll look into adding a separate benchmark with the different perf calculation in a subsequent PR.

tt-aho committed:
#0: Bump TG nightly timeout

tt-aho commented on ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_async/device/kernels/interleaved_dim3_1_1_32_any_writer.cpp:
```diff
@@ -73,11 +73,11 @@ void kernel_main() {
-    DPRINT << "fabric_connection arg 2" << get_arg_val<uint32_t>(arg_for_fab++) << "\n";
-    DPRINT << "fabric_connection arg 3" << get_arg_val<uint32_t>(arg_for_fab++) << "\n";
-    DPRINT << "fabric_connection arg 4" << get_arg_val<uint32_t>(arg_for_fab++) << "\n";
+    DPRINT << "fabric_connection arg 0 " << get_arg_val<uint32_t>(arg_for_fab++) << "\n";
```
Done

tt-aho committed:
#18798: Change ring_topology boolean flag to use Topology enum

tt-aho committed:
#18798: Add ring topology support to all gather

tt-aho committed:
#18798: Force downstream_edm_noc_interfaces array to be sized according to actually used channels
This was causing a perf drop in the mcast microbenchmark. This isn't the cleanest solution and ideally we'd make a change across all the arrays for consistency as cleanup
